app/models/log_ticket.rb

```diff
@@ -0,0 +1,27 @@
+class LogTicket < ActiveRecord::Base
+  enum backend: [ :s3, :local ]
+
+  scope :latest_pending, -> { limit(1).lock(true).select("id").where(status: "pending") }
+
+  def self.pop(key = nil, directory = nil)
+    scope = latest_pending
+    scope = scope.where(key: key) if key
+    scope = scope.where(directory: directory) if directory
+    sql = scope.to_sql
+
+    find_by_sql(["UPDATE #{quoted_table_name} SET status = ? WHERE id IN (#{sql}) RETURNING *", 'processing']).first
+  end
+
+  def filesystem
+    @fs ||=
+      if s3?
```

Comment:
@arthurnn FYI, you could stub out the S3 responses to use a local log file in tests. That's what I did in the [FastlyLogProcessor unit tests](https://github.com/rubygems/rubygems.org/blob/e000c27e764baf37f3e381b3f7e215a28cd0be0e/test/unit/fastly_log_processor_test.rb#L19-L21).

Here's a [good blog post all about stubbing AWS responses](http://ruby.awsblog.com/post/Tx15V81MLPR8D73/Client-Response-Stubs).

That seems simpler than supporting pluggable `RubygemFS` backends for LogTicket.